merge user script - fix #1472, fix #1473, fix #1481

dnil opened the pull request:
This PR adds a functionality that will winter clean the db, merging old ObjectId _id users with existing email _id users, or creating new such users, and assigning their ACMG, ClinVar, and general event entries to the new merged user.

**How to test**:
1. dry run on local, stage and prod and see that everything looks kosher.
1. live run local
1. dump relevant tables on scout-stage, then run live and evaluate.

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [x] tests executed by dNil

dnil commented:
```dry_run = True``` is there by design until done. Remove at your own peril.

dnil commented:
Local dry-run seems to work.
```
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO Running scout version 4.8.0
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO Setting database name to scout
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local scout.adapter.utils[15493] INFO Test connection with uri: mongodb://localhost:27017
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO Test if mongod is running
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local scout.adapter.client[15493] INFO Try to connect to mongodb://localhost:27017
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local scout.adapter.client[15493] INFO Connection established
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO Setting up a mongo adapter
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO ===USER===
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user: {'_id': ObjectId('5dc68fc6c77412043287df23'), 'email': 'clark.kent@mail.com', 'name': 'Superman', 'roles': ['admin', 'mme_submitter'], 'institutes': ['cust000'], 'accessed_at': datetime.datetime(2019, 11, 11, 16, 43, 23, 32000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO alt user: {'_id': 'clark.kent@mail.com', 'email': 'clark.kent@mail.com', 'name': 'Clark Kent', 'roles': ['admin', 'mme_submitter'], 'institutes': ['cust000'], 'accessed_at': datetime.datetime(2019, 11, 9, 11, 49, 9, 169000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO merged institutes: {'cust000'}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO merged roles: {'admin', 'mme_submitter'}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO update user: {'_id': 'clark.kent@mail.com', 'email': 'clark.kent@mail.com', 'name': 'Clark Kent', 'roles': ['admin', 'mme_submitter'], 'institutes': ['cust000'], 'accessed_at': datetime.datetime(2019, 11, 11, 16, 43, 23, 32000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO searching for events for user id 5dc68fc6c77412043287df23
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO ===EVENTS===
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc696c124c9340df44d3119'), 'institute': 'cust000', 'case': 'internal_id_2', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': 'http://localhost:5000/cust000/643594b/4a3a75e4287c90d108a34975253e8b93', 'category': 'variant', 'verb': 'acmg', 'subject': '1_36031420_C_T_clinical', 'level': 'specific', 'variant_id': 'f79166ee8351b94225306e31ff0d41a9', 'content': None, 'panel': None, 'created_at': datetime.datetime(2019, 11, 9, 11, 36, 49, 155000), 'updated_at': datetime.datetime(2019, 11, 9, 11, 36, 49, 155000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc696de24c9340df44d311a'), 'institute': 'cust000', 'case': 'internal_id_2', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': '/cust000/643594b/c08195419636df4efa2f837d871a679f', 'category': 'variant', 'verb': 'pin', 'subject': '1_151234968_C_T_clinical', 'level': 'specific', 'variant_id': '08194ddc8de1ee4633397a2daa2a23c8', 'content': None, 'panel': None, 'created_at': datetime.datetime(2019, 11, 9, 11, 37, 18, 676000), 'updated_at': datetime.datetime(2019, 11, 9, 11, 37, 18, 676000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc696de24c9340df44d311b'), 'institute': 'cust000', 'case': 'internal_id_2', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': '/cust000/643594b/c08195419636df4efa2f837d871a679f', 'category': 'case', 'verb': 'pin', 'subject': '1_151234968_C_T_clinical', 'level': 'specific', 'variant_id': '08194ddc8de1ee4633397a2daa2a23c8', 'content': None, 'panel': None, 'created_at': datetime.datetime(2019, 11, 9, 11, 37, 18, 677000), 'updated_at': datetime.datetime(2019, 11, 9, 11, 37, 18, 677000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc980e124c9343b4efa4c10'), 'institute': 'cust000', 'case': 'clinvar', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': '/cust000/clinvar/a401e03ef8051f3228d55a4fbe21182d', 'category': 'variant', 'verb': 'pin', 'subject': '1_156146640_G_A_clinical', 'level': 'specific', 'variant_id': '699c94368c42c704b5a5632dca37fe8e', 'content': None, 'panel': None, 'created_at': datetime.datetime(2019, 11, 11, 16, 40, 17, 733000), 'updated_at': datetime.datetime(2019, 11, 11, 16, 40, 17, 733000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc980e124c9343b4efa4c11'), 'institute': 'cust000', 'case': 'clinvar', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': '/cust000/clinvar/a401e03ef8051f3228d55a4fbe21182d', 'category': 'case', 'verb': 'pin', 'subject': '1_156146640_G_A_clinical', 'level': 'specific', 'variant_id': '699c94368c42c704b5a5632dca37fe8e', 'content': None, 'panel': None, 'created_at': datetime.datetime(2019, 11, 11, 16, 40, 17, 734000), 'updated_at': datetime.datetime(2019, 11, 11, 16, 40, 17, 734000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc980ec24c9343b4efa4c12'), 'institute': 'cust000', 'case': 'clinvar', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': '/cust000/clinvar', 'category': 'case', 'verb': 'add_phenotype', 'subject': 'clinvar', 'level': 'specific', 'variant_id': None, 'content': 'HP:0012759', 'panel': None, 'created_at': datetime.datetime(2019, 11, 11, 16, 40, 28, 855000), 'updated_at': datetime.datetime(2019, 11, 11, 16, 40, 28, 855000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc981b624c9343b4efa4c13'), 'institute': 'cust000', 'case': 'internal_id_2', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': '/cust000/643594b/2a59ad257054d249c6660c54c336802b', 'category': 'variant', 'verb': 'pin', 'subject': '1_156146640_G_A_clinical', 'level': 'specific', 'variant_id': '699c94368c42c704b5a5632dca37fe8e', 'content': None, 'panel': None, 'created_at': datetime.datetime(2019, 11, 11, 16, 43, 50, 129000), 'updated_at': datetime.datetime(2019, 11, 11, 16, 43, 50, 129000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user event: {'_id': ObjectId('5dc981b624c9343b4efa4c14'), 'institute': 'cust000', 'case': 'internal_id_2', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'link': '/cust000/643594b/2a59ad257054d249c6660c54c336802b', 'category': 'case', 'verb': 'pin', 'subject': '1_156146640_G_A_clinical', 'level': 'specific', 'variant_id': '699c94368c42c704b5a5632dca37fe8e', 'content': None, 'panel': None, 'created_at': datetime.datetime(2019, 11, 11, 16, 43, 50, 130000), 'updated_at': datetime.datetime(2019, 11, 11, 16, 43, 50, 130000)}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO searching for acmg for user id 5dc68fc6c77412043287df23
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO ===ACMG===
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO acmg: {'_id': ObjectId('5dc696c124c9340df44d3118'), 'variant_specific': '4a3a75e4287c90d108a34975253e8b93', 'variant_id': 'f79166ee8351b94225306e31ff0d41a9', 'institute_id': 'cust000', 'case_id': 'internal_id_2', 'classification': 'likely_benign', 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'user_name': 'Superman', 'created_at': datetime.datetime(2019, 11, 9, 11, 36, 49, 104000), 'criteria': []}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO searching for ClinVar for user id 5dc68fc6c77412043287df23
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO === ClinVar ===
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO acmg: {'_id': '643594_3eecfca5efea445eec6c19a53299043b', 'csv_type': 'variant', 'case_id': '643594', 'category': 'snv', 'local_id': '3eecfca5efea445eec6c19a53299043b', 'linking_id': '3eecfca5efea445eec6c19a53299043b', 'gene_symbol': 'POT1', 'ref_seq': 'NM_001042594', 'hgvs': 'c.510G>T', 'chromosome': '7', 'start': '124491972', 'stop': '124491972', 'ref': 'C', 'alt': 'A', 'variations_ids': 'rs116916706', 'condition_id_type': 'HPO', 'condition_id_value': 'HP:0002664;HP:0000014;HP:0001222;HP:0009733;HP:0000510;HP:0001251;HP:0000118;HP:0100022', 'clinsig': 'Likely benign', 'last_evaluated': '2019-06-26', 'assertion_method': 'ACMG Guidelines, 2015', 'inheritance_mode': 'Autosomal recessive inheritance', 'user': ObjectId('5dc68fc6c77412043287df23')}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO searching for clinvar submissions for user id 5dc68fc6c77412043287df23
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO === ClinVar submission ===
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO acmg: {'_id': ObjectId('5d131ffbc458940bf887a8a9'), 'status': 'open', 'created_at': datetime.datetime(2019, 6, 26, 9, 34, 19, 43000), 'user_id': ObjectId('5dc68fc6c77412043287df23'), 'institute_id': 'cust000', 'variant_data': ['643594_3eecfca5efea445eec6c19a53299043b', 'clinvar_a401e03ef8051f3228d55a4fbe21182d', 'internal_id_2_2a59ad257054d249c6660c54c336802b'], 'updated_at': datetime.datetime(2019, 11, 11, 16, 44, 14, 139000), 'case_data': ['clinvar_a401e03ef8051f3228d55a4fbe21182d_ADM1059A2', 'internal_id_2_2a59ad257054d249c6660c54c336802b_NA12882']}
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO event requests to execute: [UpdateOne({'_id': ObjectId('5dc696c124c9340df44d3119')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc696de24c9340df44d311a')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc696de24c9340df44d311b')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc980e124c9343b4efa4c10')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc980e124c9343b4efa4c11')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc980ec24c9343b4efa4c12')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc981b624c9343b4efa4c13')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc981b624c9343b4efa4c14')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None)]
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO acmg requests to execute: [UpdateOne({'_id': ObjectId('5dc696c124c9340df44d3118')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None)]
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO clinvar requests to execute: [UpdateOne({'_id': ObjectId('5dc696c124c9340df44d3119')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc696de24c9340df44d311a')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc696de24c9340df44d311b')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc980e124c9343b4efa4c10')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc980e124c9343b4efa4c11')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc980ec24c9343b4efa4c12')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc981b624c9343b4efa4c13')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None), UpdateOne({'_id': ObjectId('5dc981b624c9343b4efa4c14')}, {'$set': {'user_id': 'clark.kent@mail.com', 'user_name': 'Clark Kent'}}, False, None, None)]
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO clinvar sub requests to execute: [UpdateOne({'_id': ObjectId('5d131ffbc458940bf887a8a9')}, {'$set': {'user_id': 'clark.kent@mail.com'}}, False, None, None)]
2019-11-11 17:54:31 k1315C02VR0WRHTDH.local __main__[15493] INFO user requests to execute: [UpdateOne({'_id': 'clark.kent@mail.com'}, {'$set': {'email': 'clark.kent@mail.com', 'name': 'Clark Kent', 'roles': ['admin', 'mme_submitter'], 'institutes': ['cust000'], 'accessed_at': datetime.datetime(2019, 11, 11, 16, 43, 23, 32000)}}, False, None, None), DeleteOne({'_id': ObjectId('5dc68fc6c77412043287df23')}, None)]
```

northwestwitch commented:
This is a scary one!

northwestwitch commented:
I'll take a look!

dnil commented:
> I haven't tested it yet but you did @dnil and I think that the script will work. But I wonder if it isn't easier to just remove users with _id = ObjectID.
> 
> You have spent more time on this matter so why did you reach the conclusion that this merge is better than purge?

I'd like to save roles, dates and institutes from the OI objects. The alternative would be to look through all the users manually to make sure it is safe to delete them and their potential info?

dnil commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+    event_requests = []
+    acmg_requests = []
+    clinvar_requests = []
+    clinvar_submission_requests = []
```
I presume so: they have an event, and no collection that I am aware of?

dnil commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+    clinvar_submission_requests = []
+
+    for oi_user_obj in adapter.user_collection.find({'_id': {'$type':'objectId'}}):
+        if not oi_user_obj.get('email'):
```
Yes, at least one; but that one does not seem to connect to any events fortunately. It seems it would have been a little hard to log in with any decently modern version of the codebase, so probably predates events.

dnil commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+        LOG.info("user: {}".format(oi_user_obj))
+
+        create_alt = False
+        alt_user_obj = adapter.user_collection.find_one({'_id': oi_user_email})
```
Well, that is the other user. Normally the one they have been using for the last year(s) or so, at least up until the ldap change.

northwestwitch replied:
So the stats and events associated to these old double accounts are not taken into account today. Erasing the old users wouldn't change the data we were seeing before the bug. Nonetheless it would be nice to integrate the old users stats and events so we would have complete info for the users from now on.

dnil replied:
Not **much** no; but then there was this issue of last week and today, with 100+ edits to cater for. But e.g. when an account was first created for a user might be nice to save. Just disabling the "old" account and re-routing events etc to the new one would have worked as well.

dnil commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+
+        create_alt = False
+        alt_user_obj = adapter.user_collection.find_one({'_id': oi_user_email})
+        if not alt_user_obj:
```
I will agree that it is what caused the current spur of logins to the old accounts, but they have been around for a long time, and they have been used way-back-when as well.

Also, since we have had several users use their ObjectId over a few days now (e.g. Bianca, Sofie, Per, me..) there will be quite a few recent edits that would crash stuff/lack user names and other awkwardness if we just drop the users. 

So we do need to **at least** connect those events etc to their corresponding _id == email users.

dnil replied:
Perhaps if I tell you I could not dry run on the prod db without import copy?

northwestwitch replied:
> So we do need to **at least** connect those events etc to their corresponding _id == email users.

I agree

northwestwitch replied:
> Perhaps if I tell you I could not dry run on the prod db without import copy, and that the following block is the only place I use it?

I don't understand, what does it entail?

dnil replied:
Simply that it is being used when fed with live data, but only as edge cases.

dnil commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+        user_requests.append(operation)
+
+        ###
+        ### events
```
Good question! At least the bulk is from today, but with a tail from last weeks attempt(s) with 4.8 as well as today. Then a small number of users with really old ones as well; think Robin..

northwestwitch replied:
Ok!

northwestwitch commented:
> I'd like to save roles, dates and institutes from the OI objects. The alternative would be to look through all the users manually to make sure it is safe to delete them and their potential info? In the end, there are apparently rather small deviations between the two approaches.

One could also check when it was their last login. Like if it was in 2017 and the other user was last week.. I think role and institutes will be actual if from user with _id = email.

dnil commented:
> > I'd like to save roles, dates and institutes from the OI objects. The alternative would be to look through all the users manually to make sure it is safe to delete them and their potential info? In the end, there are apparently rather small deviations between the two approaches.
> 
> One could also check when it was their last login. Like if it was in 2017 and the other user was last week.. I think role and institutes will be actual if from user with _id = email.

That is likely correct; I thought it was more mixed this morning. The deviations from _id=email being correct on role and institute is likely small, and could be considered potential noise/outdated info.

northwestwitch commented:
Running this script is fine by me. As a precaution one could also backup all the involved collections of prod database (it wouldn't occupy too much space I guess?).

dnil commented:
> As a precaution one could also backup all the involved collections of prod database (it wouldn't occupy too much space I guess?).
I agree: I thought the same, blessed the design that didn't put user stuff down in variant_collection
but haven't got around to checking docks for how to dump selected collections.

moonso commented:
It is possible to run: 
```
mongodump --port 27017 --db=scout --collection=User --out dump/
```
More information here: https://docs.mongodb.com/manual/tutorial/backup-and-restore-tools/

dnil commented:
Thanks! I tried it on relevant collection on scout-stage; seems to work! Have you found any other collection or user_id somewhere that needs changing?

moonso commented:
I'm investigating now...

Meanwhile, I wonder if you could add to the docstring just a few lines what the script will do?
Also I wonder: Why not delete the users that we do not want?

dnil commented:
> Meanwhile, I wonder if you could add to the docstring just a few lines what the script will do?
> Also I wonder: Why not delete the users that we do not want?

Sure, will do! I tried to answer previously in the thread. 

This does delete the users with ObjectId _id. It only makes sure to preserve the info from them on their "main" account in the process - or create a new user if need be.

moonso commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+        else:
+            LOG.info("alt user: {}".format(alt_user_obj))
+            merged_institutes = set()
+            merged_institutes.update(alt_user_obj.get('institutes') + oi_user_obj.get('institutes'))
```
```suggestion
            merged_institutes.update(alt_user_obj.get('institutes',[]) + oi_user_obj.get('institutes', []))
```
This will fail if institutes is None or does not exist

dnil replied:
Same as below; I'll buy it since it is at least as good as before, and we can argue semantics of edge cases later. 😈

moonso commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+            alt_user_obj['institutes'] = list(merged_institutes)
+
+            merged_roles = set()
+            merged_roles.update(alt_user_obj.get('roles') + oi_user_obj.get('roles'))
```
```suggestion
            merged_roles.update(alt_user_obj.get('roles',[]) + oi_user_obj.get('roles',[]))
```

dnil replied:
Hm, wouldn't None and [] just be equivalent here? Oh well, this is even more secure so I will go with this.

moonso commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+
+            created_at = oi_user_obj.get('created_at')
+            alt_created_at = alt_user_obj.get('created_at')
+            if (alt_created_at and alt_created_at < created_at):
```
```suggestion
            if ((alt_created_at and created_at) and alt_created_at < created_at):
```

dnil replied:
Since there is a check for existence of created_at afterwards, and None is smaller than all valid alt_created_at I dont think it matters - but also won't really hurt, so adding!

moonso replied:
```
In [5]: from datetime import datetime

In [6]: a = datetime.now()

In [7]: a < None
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-cfa6709df3e1> in <module>()
----> 1 a < None

TypeError: unorderable types: datetime.datetime() < NoneType()
```

dnil replied:
I stand corrected! 🥇

moonso commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+
+            accessed_at = alt_user_obj.get('accessed_at')
+            oi_accessed_at = oi_user_obj.get('accessed_at')
+            if (oi_accessed_at and oi_accessed_at > accessed_at):
```
```suggestion
            if ((oi_accessed_at and accessed_at) and oi_accessed_at > accessed_at):
```

dnil replied:
:+1:

moonso commented on scripts/merge_users.py:
```diff
@@ -0,0 +1,285 @@
+
+        if create_alt:
+            LOG.info("create user: {}".format(alt_user_obj))
+            operation = pymongo.InsertOne(alt_user_obj)
```
There is no DryRun clause here!!

dnil replied:
Meh, this is just appending future operations: bulk write will happen at the end, well protected by dry run clauses. I hope.

moonso replied:
Haha sorry

dnil replied:
😅

dnil commented:
Local tested ok!
Stage, in summary also worked:
```
Modified 2 ACMG.
Modified 1 case submissions.
Modified users with the following: {'writeErrors': [], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 38, 'nModified': 6, 'nRemoved': 38, 'upserted': []}
```
